File: 0352_dataStreamAsDisjointIntervals/main.py
from typing import List
from sortedcontainers import SortedList
class SummaryRanges:

    def __init__(self):
        self.intervals = SortedList()
        

    def addNum(self, value: int) -> None:
        n = len(self.intervals)
        idx = self.intervals.bisect_right((value, float("inf")))
        
        lTree = self.intervals[idx-1] if 0 <= idx-1 < n else (float("-inf"), float("-inf"))
        rTree = self.intervals[idx] if 0 <= idx < n else (float("inf"), float("inf"))

        if lTree[0] <= value <= lTree[1]:
            return
        
        l, r = value, value
        if lTree[1] + 1 == value:
            self.intervals.remove(lTree)
            l = lTree[0]
        
        if rTree[0] - 1 == value:
            self.intervals.remove(rTree)
            r = rTree[1]
        
        self.intervals.add((l, r))

    def getIntervals(self) -> List[List[int]]:
        return self.intervals


# Your SummaryRanges object will be instantiated and called as such:
def exe(obj, method, value):
    if method == "addNum":
        obj.addNum(value[0])
    if method == "getIntervals":
        obj.getIntervals()

obj = SummaryRanges()
method = ["SummaryRanges","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals","addNum","getIntervals"]
value = [[],[6],[],[6],[],[0],[],[4],[],[8],[],[7],[],[6],[],[4],[],[7],[],[5],[]]
for i in range(1, len(method)):
    exe(obj, method[i], value[i])
    print(obj.intervals)

File: 0352_dataStreamAsDisjointIntervals/main.cpp
#include <vector>
#include <map>

class SummaryRanges {
private:
    std::map<int, std::vector<int>> intervals;

public:
    SummaryRanges() {
        
    }
    
    void addNum(int value) {
        std::map<int, std::vector<int>>::iterator rIt = intervals.upper_bound(value); // first element that > value
        std::map<int, std::vector<int>>::iterator lIt = intervals.end();
        if(rIt != intervals.begin()) {
            lIt = rIt;
            --lIt;
        }

        if(lIt != intervals.end() && value >= (*lIt).second[0] && value <= (*lIt).second[1])
            return;
        
        int l = value;
        int r = value;
        if(lIt != intervals.end() && (*lIt).second[1] + 1 == value) {
            l = (*lIt).second[0];
            intervals.erase(lIt);
        }

        if(rIt != intervals.end() && (*rIt).second[0] - 1 == value) {
            r = (*rIt).second[1];
            intervals.erase(rIt);
        }
        intervals[l] = {l, r};
    }
    
    std::vector<std::vector<int>> getIntervals() {
        std::vector<std::vector<int>> res;
        for(const auto& [_, interval]: intervals)
            res.push_back(interval);
        return res;
    }
};

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges* obj = new SummaryRanges();
 * obj->addNum(value);
 * vector<vector<int>> param_2 = obj->getIntervals();
 */